### PR TITLE
fix(walreceiver): Timeline::shutdown can leave a dangling handle_walreceiver_connection tokio task

### LINKED ISSUE #7062: Timeline::shutdown can leave a dangling `handle_walreceiver_connection` tokio task
# Problem

```
diff --git a/pageserver/src/tenant/timeline/walreceiver.rs b/pageserver/src/tenant/timeline/walreceiver.rs
index 2fab6722b..2a2f9ab04 100644
--- a/pageserver/src/tenant/timeline/walreceiver.rs
+++ b/pageserver/src/tenant/timeline/walreceiver.rs
@@ -236,6 +236,10 @@ impl<E: Clone> TaskHandle<E> {
     async fn shutdown(self) {
         if let Some(jh) = self.join_handle {
             self.cancellation.cancel();
+            // we can stop being polled before the jh completes.
+            // it can happen because connection_manager_loop_step is in a tokio::select!() statement together with task_mgr cancellation:
+            // https://github.com/neondatabase/neon/blob/fb518aea0db046817987a463b1556ad950e97f09/pageserver/src/tenant/timeline/walreceiver.rs#L84-L104
+            // the consequence of that will be that we leave a tokio task that executes handle_walreceiver_connection running
             match jh.await {
                 Ok(Ok(())) => debug!("Shutdown success"),
                 Ok(Err(e)) => error!("Shutdown task error: {e:?}"),
```

Found together with @koivunej  while investigating https://github.com/neondatabase/neon/pull/7051

i.e., rapid `/location_config` that transition a tenant from Single to Multi and back with changing generation numbers, so, the code takes the slow path.

Consequence of this:

- the dangling task can continue to ingest WAL
- that means it can continue to try to
  - `put` / `finish_write` into the Timeline
    - which in turn can cause layers to be frozen & flushed
  - trigger on-demand downloads

We audited the codebase to ensure that, as of #7051 , none of the above will succeed after `Timeline::shutdown` returns.

# Solution Proposal

Don't cancel-by-drop the `connection_manager_loop_step`.
Use `shutdown_token()` instead and make code responsive to it.

# Impl

```[tasklist]
### Tasks
- [ ] https://github.com/neondatabase/neon/pull/7234
- [ ] https://github.com/neondatabase/neon/pull/7235
- [ ] https://github.com/neondatabase/neon/pull/7260
- [ ] https://github.com/neondatabase/neon/pull/7233
```

### PR DESCRIPTION
**Pull request by problame**
# Problem

As pointed out through doc-comments in this PR, `drop_old_connection` is not cancellation-safe.

This means we can leave a `handle_walreceiver_connection` tokio task dangling during Timeline shutdown.

More details described in the corresponding issue #7062.

# Solution

Don't cancel-by-drop the `connection_manager_loop_step` from the `tokio::select!()` in the task_mgr task.
Instead, transform the code to use a `CancellationToken` --- specifically, `task_mgr::shutdown_token()` --- and make code responsive to it.

The `drop_old_connection()` is still not cancellation-safe and also doesn't get a cancellation token, because there's no point inside the function where we could return early if cancellation were requested using a token.

We rely on the `handle_walreceiver_connection` to be sensitive to the `TaskHandle`s cancellation token (argument name: `cancellation`).
Currently it checks for `cancellation` on each WAL message. It is probably also sensitive to `Timeline::cancel` because ultimately all that `handle_walreceiver_connection` does is interact with the `Timeline`.

In summary, the above means that the following code (which is found in `Timeline::shutdown`) now might **take longer**, but actually ensures that all `handle_walreceiver_connection` tasks are finished:

```rust
task_mgr::shutdown_tasks(
    Some(TaskKind::WalReceiverManager),
    Some(self.tenant_shard_id),
    Some(self.timeline_id)
)
```

# Refs

refs #7062

### CODE REVIEW

File: pageserver/src/tenant/timeline/walreceiver/connection_manager.rs [hunk 1]
```diff
@@ -127,13 +138,14 @@ pub(super) async fn connection_manager_loop_step(
...
                         }
+                        // TODO: can this use the `cancel` token?
                         connection_manager_state.drop_old_connection(false).await;
                     },
                 }
             },
 
             // Got a new update from the broker
-            broker_update = broker_subscription.message() => {
+            broker_update = broker_subscription.message() /* TODO: review cancellation-safety */ => {
```
Comment (problame) [inline pageserver/src/tenant/timeline/walreceiver/connection_manager.rs:148]: We were relying on cancellation-safety before this PR, so, this shouldn't be a blocker.

File: pageserver/src/tenant/timeline/walreceiver/connection_manager.rs [hunk 2]
```diff
@@ -241,9 +253,14 @@ async fn subscribe_for_timeline_updates(
             subscription_key: Some(key),
         };
 
-        match broker_client.subscribe_safekeeper_info(request).await {
+        match {
+            tokio::select! {
+                r = broker_client.subscribe_safekeeper_info(request) => { r }
+                _ = cancel.cancelled() => { return Err(Cancelled); }
+            }
+        } {
```
Comment (problame) [inline pageserver/src/tenant/timeline/walreceiver/connection_manager.rs:261]: Is the gRPC generated client method cancel-safe?

Even before this PR, we did implicitly rely on cancel-safety because the surrounding `subscribe_for_timeline_updates` could have been cancelled further upstack.

But, I don't think anyone really checked cancel-safety before this PR.

So, would be nice to have a definitive answer & add a comment explaining that we checked for cancel-safety of the generated method here.

Comment (koivunej) [reply pageserver/src/tenant/timeline/walreceiver/connection_manager.rs:260]: I would expect it to be fine as it would look like from issues the h2 level is used for cancelling, but will check...

Comment (koivunej) [reply pageserver/src/tenant/timeline/walreceiver/connection_manager.rs:260]: There is an mpsc channel in between so cancellation will be communicated via it cleanly: https://github.com/neondatabase/neon/blob/12512f31736a5c5b3d3973c5c5cfd43dd58acb3d/storage_broker/src/lib.rs#L40

**Review by koivunej — APPROVED** [review]
Running the connection manager loop step to completion and not racing to cancellation token, but ensuring it is done internally within each step, which this PR adds. `tonic` stuff seems cancellation safe at least by construction, didn't look if a mpsc channel drop is handled in tower buffer, but I trust it is.

**Comment by koivunej** [comment]
I would prefer to see the subject for this PR and commit be:

```
-fix(#7062): ...
+fix(walreceiver): ...
```

The used form looks like conventional commits, which I often try to use as well. There should be some sort of component between the parentheses, not an issue number.
